src/uu/ls/src/ls.rs

```diff
@@ -1510,27 +1578,54 @@ fn enter_directory(dir: &PathData, config: &Config, out: &mut BufWriter<Stdout>)
             }
         };
 
-        if should_display(&unwrapped, config) {
-            // why check the DirEntry file_type()?  B/c the call is
-            // nearly free compared to a metadata() or file_type() call on a dir/file
-            let path_data = match unwrapped.file_type() {
-                Err(err) => {
-                    let _ = out.flush();
-                    show!(LsError::IOErrorContext(err, unwrapped.path()));
-                    continue;
-                }
-                Ok(dir_ft) => {
-                    PathData::new(unwrapped.path(), Some(Ok(dir_ft)), None, config, false)
+        if should_display(&dir_entry, config) {
+            // Why prefer to check the DirEntry file_type()?  B/c the call is
+            // nearly free compared to a metadata() or file_type() call on a dir/file.
+            //
+            // Why not print an error here?  If we wait for the metadata() call, we make
+            // certain we print the error once.  This also seems to match GNU behavior.
+            let entry_path_data = match dir_entry.file_type() {
```

Comment:
I would move the logic of this `if` statement into `PathData` as follows:
- Here we pass `None` for both `metadata` and `file_type`.
- The metadata is already retrieved from the `DirEntry` when we call `md()`, so the correct metadata will still be used.
- The `ft()` function then can be adapted to uses `DirEntry::file_type()` if `de` is available.

Reply:
Additionally, if `None` is passed here, I think the `metadata` and `file_type` parameters can be removed from the `PathData::new` function, because in all other locations they're `None` as well.

Reply:
Makes sense.